src/kcas.ml

```diff
@@ -8,181 +8,167 @@ module type Backoff = Backoff.S
-      rdcss_read a
-  | _ -> r
-
-let rec casn_proceed c =
-  let rec phase1 curr_c_l curr_st out =
-    match curr_c_l with
-    | CAS (a, o, n) :: curr_c_t_tail when curr_st = SUCCEEDED -> (
-        let s = rdcss (mk_rdcss c.st (WORD UNDECIDED) a o (CASN_DESC c)) in
-        match s with
-        | CASN_DESC c' ->
-            if c.id_casn != c'.id_casn then (
-              ignore @@ casn_proceed c' ;
-              phase1 curr_c_l curr_st out )
-            else phase1 curr_c_t_tail curr_st (CAS (a, o, n) :: out)
-        | _ ->
-            if st_eq s o then
-              phase1 curr_c_t_tail curr_st (CAS (a, o, n) :: out)
-            else phase1 curr_c_t_tail FAILED (CAS (a, o, n) :: out) )
-    | _ ->
-        ignore @@ commit (CAS (c.st, WORD UNDECIDED, WORD curr_st)) ;
-        out
-  in
-  let rec phase2 curr_c_l succ =
-    match curr_c_l with
-    | CAS (a, o, n) :: curr_c_l_tail -> (
-      match Atomic.get succ with
-      | WORD SUCCEEDED ->
-          ignore @@ commit (CAS (a, CASN_DESC c, n)) ;
-          phase2 curr_c_l_tail succ
-      | _ ->
-          ignore @@ commit (CAS (a, CASN_DESC c, o)) ;
-          phase2 curr_c_l_tail succ )
-    | _ -> Atomic.get succ = WORD SUCCEEDED
-  in
-  match Atomic.get c.st.content with
-  | WORD UNDECIDED -> phase2 (phase1 c.c_l SUCCEEDED []) c.st.content
-  | _ -> phase2 c.c_l c.st.content
-
-let rec get a =
-  let r = rdcss_read a.content in
-  match r with
-  | CASN_DESC c ->
-      ignore @@ casn_proceed c ;
-      get a
-  | WORD out -> out
-  | _ -> assert false
-
-let kCAS c_l =
-  match c_l with
+let set atom value =
+  Atomic.set atom.state { before = value; after = value; casn = casn_after }
+
+let rec release_after = function
+  | [] -> true
+  | CAS (_, cas) :: cass ->
+      cas.casn <- casn_after;
+      cas.before <- cas.after;
+      release_after cass
+
+let rec release_before = function
+  | [] -> false
+  | CAS (_, cas) :: cass ->
+      cas.casn <- casn_before;
+      cas.after <- cas.before;
+      release_before cass
+
+(* Note: The writes to `cas.casn <- ...` above could be removed to reduce
+   time at the cost of increasing space usage (by a constant factor). *)
+
+let release cass = function
+  | `After -> release_after cass
+  | `Before -> release_before cass
+
+let finish casn (`Undetermined cass as undetermined)
+    (status : [ `Before | `After ]) =
+  if Atomic.compare_and_set casn (undetermined :> status) (status :> status)
+  then release cass status
+  else Atomic.get casn == `After
+
+let rec determine casn undetermined = function
+  | CAS (atom_state, state) :: cass_1 as cass_0 ->
+      let state' = Atomic.get atom_state in
+      if state == state' then determine casn undetermined cass_1
+      else
+        let before = state.before in
+        let before' = state'.before and after' = state'.after in
+        if
+          (before == before' && before == after')
+          || before == if is_after state'.casn then after' else before'
```

Comment:
This logic of checking whether previous operation finished and getting its result happens here, in `cas` and `get`. Reusing `get` would make it more readable.

Reply:
Yes. I believe I wrote it that way for minor performance reasons.  I'll try and see if I can reduce the duplication without giving up performance.

* [ ] check and report on whether duplication can be reduced

Reply:
Hmm... It seems to get tricky to avoid the duplication with or without performance degradation.

First, the existential poses typing challenges.  Moving `get` into the same `let rec` and calling it results in:

```
File "src/kcas.ml", line 78, characters 34-38:
78 |       else if state.before == get atom then
                                       ^^^^
Error: This expression has type $CAS_'a ref
       but an expression was expected of type 'a
       The type constructor $CAS_'a would escape its scope
```

This tends to happen when trying to extract pretty much any useful part out of the loop.

Second, there are some intricate choices made in the comparisons for performance.  In particular, calls to `is_after` and accessing the `casn` is delayed as much as possible. The seemingly redundant comparisons before the `is_after` call improve performance by not accessing the `casn` unless necessary and those comparisons slightly vary depending on what is at hand.  In particular,